iavl_nodedb.go

```diff
@@ -73,68 +107,418 @@ func (ndb *nodeDB) SaveNode(node *IAVLNode) {
+
+		// See comment on `orphansKeyFmt`. Note that here, `version` and
+		// `toVersion` are always equal.
+		fmt.Sscanf(string(key), orphansKeyFmt, &toVersion, &fromVersion)
+
+		// Delete orphan key and reverse-lookup key.
+		ndb.batch.Delete(key)
+		ndb.batch.Delete([]byte(fmt.Sprintf(orphansIndexKeyFmt, hash)))
+
+		// If there is no predecessor, or the predecessor is earlier than the
+		// beginning of the lifetime (ie: negative lifetime), or the lifetime
+		// spans a single version and that version is the one being deleted, we
+		// can delete the orphan.  Otherwise, we shorten its lifetime, by
+		// moving its endpoint to the previous version.
+		if predecessor < fromVersion || fromVersion == toVersion {
+			ndb.batch.Delete(ndb.nodeKey(hash))
+			ndb.uncacheNode(hash)
+		} else {
+			ndb.saveOrphan(hash, fromVersion, predecessor)
+		}
+	})
+}
+
+func (ndb *nodeDB) nodeKey(hash []byte) []byte {
+	return []byte(fmt.Sprintf(nodesKeyFmt, hash))
+}
+
+func (ndb *nodeDB) getLatestVersion() uint64 {
+	if ndb.latestVersion == 0 {
+		ndb.getVersions()
+	}
+	return ndb.latestVersion
+}
+
+func (ndb *nodeDB) getVersions() map[uint64][]byte {
+	if len(ndb.versionCache) == 0 {
+		ndb.traversePrefix([]byte(rootsPrefix), func(k, hash []byte) {
+			var version uint64
+			fmt.Sscanf(string(k), rootsPrefixFmt, &version)
+			ndb.cacheVersion(version, hash)
+		})
+	}
+	return ndb.versionCache
+}
+
+func (ndb *nodeDB) cacheVersion(version uint64, hash []byte) {
+	ndb.versionCache[version] = hash
+
+	if version > ndb.getLatestVersion() {
+		ndb.latestVersion = version
+	}
+}
+
+func (ndb *nodeDB) getPreviousVersion(version uint64) uint64 {
+	var result uint64 = 0
+	for v, _ := range ndb.getVersions() {
+		if v < version && (result == 0 || v > result) {
+			result = v
+		}
+	}
+	return result
+}
+
+// deleteRoot deletes the root entry from disk, but not the node it points to.
+func (ndb *nodeDB) deleteRoot(version uint64) {
+	key := fmt.Sprintf(rootsPrefixFmt, version)
+	ndb.batch.Delete([]byte(key))
+
+	delete(ndb.versionCache, version)
+
+	if version == ndb.getLatestVersion() {
+		cmn.PanicSanity("Tried to delete latest version")
 	}
-	if !node.persisted {
-		cmn.PanicSanity("Shouldn't be calling remove on a non-persisted node.")
+}
+
+func (ndb *nodeDB) traverseOrphans(fn func(k, v []byte)) {
+	ndb.traversePrefix([]byte(orphansPrefix), fn)
+}
+
+// Traverse orphans ending at a certain version.
+func (ndb *nodeDB) traverseOrphansVersion(version uint64, fn func(k, v []byte)) {
+	prefix := fmt.Sprintf(orphansPrefixFmt, version)
+	ndb.traversePrefix([]byte(prefix), fn)
+}
+
+// Traverse all keys with a certain prefix.
+func (ndb *nodeDB) traversePrefix(prefix []byte, fn func(k, v []byte)) {
+	if ldb, ok := ndb.db.(*dbm.GoLevelDB); ok {
+		it := ldb.DB().NewIterator(util.BytesPrefix([]byte(prefix)), nil)
+		for it.Next() {
+			k := make([]byte, len(it.Key()))
+			v := make([]byte, len(it.Value()))
+
+			// Leveldb reuses the memory, we are forced to copy.
```

Comment:
Hmmm. Can we not have the individual `fn` that need to keep the values around do the copying? Seems there are cases where we don't need copies. 

Of course that wouldn't be the safest thing and would require clear comments everywhere, but just checking ...

Reply:
Yeah I think this behavior is so unexpected and low-level that I wouldn't want to expose it. I wasted a good 15 minutes trying to figure out what was wrong.